### PR TITLE
access_monitoring_rules: Add timezone to schedules spec

### PR DESCRIPTION
**Pull request by bernardjkim**
Supports: https://github.com/gravitational/teleport/issues/51682, https://github.com/gravitational/teleport/issues/55075
RFD: https://github.com/gravitational/teleport/pull/57334

This extends the access_monitoring_rules schedules spec with timezones.

This was previously omitted from https://github.com/gravitational/teleport/pull/59007 after choosing to store all schedules in the backend in UTC. However, after attempting to implement the timezone conversions on the frontend, I've realized this approach is not compatible with the frontend design for schedules.

There are a couple reasons why converting to UTC in the backend is not compatible with the design, but the main reason is that the frontend design limits 1 shift per day. This is a problem because valid schedules configured using the standard editor can easily be converted into a schedule that is not valid within the schedule editor.

For example, if a user configures the following schedule using the standard editor:
Screenshot 2025-10-08 at 2 30 04 PM

This schedule would then be converted into UTC in the backend, and stored as:
```yaml
schedules:
  default:
    time:
      shifts:
      - weekday: Monday
        start: 19:00
        end: 23:59
      - weekday: Tuesday
        start: 00:00
        end: 6:59
      - weekday: Tuesday
        start: 19:00
        end: 23:59
      - weekday: Wednesday
        start: 00:00
        end: 6:59
```

If the user attempts to edit this schedule again within the WebUI, they would not be able to use the standard editor because there are two shifts for Tuesday.

### CODE REVIEW

File: api/proto/teleport/accessmonitoringrules/v1/access_monitoring_rules.proto [hunk 1]
```diff
@@ -103,6 +103,9 @@ message TimeSchedule {
   // Shifts are configured in UTC.
   repeated Shift shifts = 1;
 
+  // Timezone specifies the schedule timezone.
+  string timezone = 2;
```
Comment (marcoandredinis) [inline api/proto/teleport/accessmonitoringrules/v1/access_monitoring_rules.proto]: This field should reference the expected format for the timezones.
Either some publicly accessible list or at least a couple of examples (ideally, the most common ones).

File: lib/services/access_monitoring_rules.go [hunk 2]
```diff
@@ -156,6 +156,14 @@ func validateSchedules(schedules map[string]*accessmonitoringrulesv1.Schedule) e
 }
 
 func validateTimeSchedule(schedule *accessmonitoringrulesv1.TimeSchedule) error {
+	if schedule.GetTimezone() == "" {
+		return trace.BadParameter("timezone is required")
+	}
+
+	if _, err := time.LoadLocation(schedule.GetTimezone()); err != nil {
```
Comment (marcoandredinis) [inline lib/services/access_monitoring_rules.go:160]: [time.LoadLocation](https://pkg.go.dev/time#LoadLocation) states that:
```go
// LoadLocation looks for the IANA Time Zone database in the following
// locations in order:
//
//   - the directory or uncompressed zip file named by the ZONEINFO environment variable
//   - on a Unix system, the system standard installation location
//   - $GOROOT/lib/time/zoneinfo.zip
//   - the time/tzdata package, if it was imported
```

Maybe it's worth testing whether the files are available in our distroless image.

File: lib/services/access_monitoring_rules.go [hunk 3]
```diff
@@ -156,6 +156,14 @@ func validateSchedules(schedules map[string]*accessmonitoringrulesv1.Schedule) e
...
+		return trace.Wrap(err, "timezone is invalid")
```
Comment (marcoandredinis) [inline lib/services/access_monitoring_rules.go]: Can we also return some reference to which values are valid?

File: lib/services/access_monitoring_rules_test.go [hunk 4]
```diff
@@ -242,18 +246,41 @@ func TestValidateSchedules(t *testing.T) {
...
-					Time: &accessmonitoringrulesv1.TimeSchedule{},
+					Time: &accessmonitoringrulesv1.TimeSchedule{
+						Timezone: time.UTC.String(),
+					},
 				},
 			},
 			assertErr: func(t require.TestingT, err error, _ ...interface{}) {
 				require.ErrorContains(t, err, "at least one shift is require")
 			},
 		},
```
Comment (marcoandredinis) [inline lib/services/access_monitoring_rules_test.go:251]: It can be useful to add tests for other timezones

**Review by marcoandredinis — COMMENTED** [review]
Honestly, I don't like the idea of storing timezones in the backend.

First, there's the timezone naming: our WebUI can help here but what about other clients?
tsh, terraform, kube operator and other teleport-api clients
Should we use the abbreviation format or `continent/city` format?
(eg, WET vs Europe/Lisbon)
Given the printscreen, we use the former, but the unit tests use the latter.
Can we use both?

Then, there's the format we support: only abbreviated like WEST, PDT, and so on?
Or also the `continent/city` format?
If using abbreviated, can we set WET and WEST?
WET only applies during half of the year, and WEST during the other half.
What happens if a shift is set to WEST timezone but we are not in that part of the year (thus, using WET).
(the same applies to PDT/PST)

Should we only accept the format where we have a specific location (continent/city like format)?

Can you please add a couple more tests to exercise these edge cases?


The fact that the UI does not allow for shifts that overlap two sequential days should not deter us from providing a great UX to the end user.

As an example, we could use something like this to have an overall visualization of the configured shifts:

Image

File: lib/services/access_monitoring_rules.go (hunk 2 above)
Comment (bernardjkim) [reply lib/services/access_monitoring_rules.go:160]: From what I can tell, it doesn't seem like the specified files are available in the distroless image. However, I have tested several timezone values in the cloud env and the values seemed to work as expected.

I'm guessing the last option is being used and I think `time/tzdata` package is being imported by a dependency. But we can explicitly import it in this file as well to be safe.

**Comment by bernardjkim** [comment]
Thanks for all the great feedback so far!

> Honestly, I don't like the idea of storing timezones in the backend.

Yeah I agree with you here. I just couldn't find a way to make this compatible with the WebUI design.

> First, there's the timezone naming: our WebUI can help here but what about other clients?

We can make the `timezone` field optional and default to `UTC`. So for non-WebUI clients, this will essentially be a no-op change.

> Should we use the abbreviation format or continent/city format?

According to the documentation, `time.LoadLocation` function accepts `UTC` and `continent/city` formatted timezone values. This should take in to consideration day light savings changes, I believe.

> The fact that the UI does not allow for shifts that overlap two sequential days should not deter us from providing a great UX to the end user. As an example, we could use something like this to have an overall visualization of the configured shifts:

The example UI design you've shared definitely makes more sense for this kind of schedule configuration. That is what I have seen in most other schedule configuration UIs. However, I think it is too late to request a UI change from the design team at this point. I think the design was created with a specific use-case in mind, and it was scoped to meet the requesting customer's requirements.

If we have an opportunity to redesign this UI, I would definitely advocate for a different design that is more like what you've posted.

**Review by marcoandredinis — APPROVED** [review]

**Review by ptgott — APPROVED** [review]
Just a reminder that if we're going to release this in a minor version after backporting, we should specify in the docs which minor version the new fields require.

**Review by r0mant — APPROVED** [review]
